Commands: Add /botmsg and inheritance to /noreply

sirDonovan opened the pull request:
This would be nice to have for command-heavy bot pages.

Zarel commented:
I'm not sure I like hiding what's going on there...

Zarel commented on server/chat-commands/core.ts:
```diff
@@ -539,7 +539,7 @@ export const commands: ChatCommands = {
 		}
 
-		return this.parse(target);
+		return this.parse(target, this.isQuiet);
```
Instead of passing `this.isQuiet` as an argument, should `this.parse` automatically inherit the `isQuiet` setting of the parent context?

sirDonovan replied:
A potential issue with that would be hiding the help text if a command were used incorrectly.

Zarel replied:
Fair enough. I think it should at least be an option argument rather than a positional argument.

sirDonovan replied:
Updated.

Zarel commented:
Although perhaps the correct answer is for `/help` and error messages to bypass `/noreply`?

Zarel commented:
I'm also somewhat afraid this would get abused to trick people into PMing other users bad things.

Some sort of prefix like `[scripted bot message]` or some check for a bot rank might be necessary here.

Zarel commented:
Perhaps a different command entirely, like `/botcmd`

mia-pi-git commented on server/chat.ts:
```diff
@@ -444,11 +444,11 @@ export class CommandContext extends MessageContext {
 			// spawn subcontext
 			const subcontext = new CommandContext(this);
-			if (quiet) subcontext.isQuiet = true;
+			if (options && options.isQuiet) subcontext.isQuiet = true;
```
```suggestion
			if (options?.isQuiet) subcontext.isQuiet = true;
```

sirDonovan commented:
`CommandContext` now inherits `isQuiet` but bypasses it for `/help` and error replies. I also added a check for bot rank when a button contains `/noreply /msg`.

Zarel commented:
I think I'd prefer `/botmsg` to `/noreply msg`. I don't think of showing the sent PM as a "reply" exactly.

sirDonovan commented:
Updated to `/botmsg` and also removed the `/noreply` bypass for errors as that's intended for things like https://github.com/smogon/pokemon-showdown-client/pull/1600

Zarel commented:
This part doesn't seem to have been implemented, and it also seems like the most important part:

> I'm also somewhat afraid this would get abused to trick people into PMing other users bad things.
> 
> Some sort of prefix like `[scripted bot message]` or some check for a bot rank might be necessary here.

The check for a bot rank is probably best.

Zarel commented:
Oh, I have a pretty simple fix for this: Require `/botmsg` to be used in the room the bot is auth of.
